packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart

```diff
@@ -0,0 +1,173 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/painting.dart';
+import 'package:flutter/scheduler.dart';
+
+import 'framework.dart';
+import 'scrollable.dart';
+
+/// Provides access to a [BuildContext] for a [ScrollAwareImageProvider].
+///
+/// Creators of this object must guarantee the following:
+///
+///   1. They create this object at or after [State.initState] but before
+///      [State.dispose]. In particular, do not attempt to create this from the
+///      constructor of a state.
+///   2. They call [dispose] from [State.dispose].
+///
+/// This object will not hold on to the [State] after disposal.
+@optionalTypeArgs
+class ScrollAwareContextProvider<T extends State> {
+  /// Creates a provider of a [BuildContext] for a [ScrollAwareImageProvider].
+  ///
+  /// Callers must call [dispose] when the [State] is disposed.
+  ///
+  /// The [State] must not be null, and [State.mounted] must be true.
+  ScrollAwareContextProvider(this._state)
+      : assert(_state != null),
+        assert(_state.mounted, 'A ScrollAwareContextProvider was given a BuildContext for an Element that was never mounted.');
+
+  /// Provides safe access to the build context.
+  ///
+  /// If [dispose] has been called, will return null.
+  ///
+  /// Otherwise, asserts the [_state] is still mounted and returns its context.
+  BuildContext get _context {
```

Comment:
making this private means that nobody outside this file can use this class. if this class is valuable enough that we're making it public, let's make it reusable as well.

Reply:
This is on purpose - if we make it reusable, it should probably go somewhere else and have a different name.  

The class has to be public because it's a public parameter to the constructor of ScrollAwareImageProvider.